### DISCUSSION TITLE
What is the Best Approach for Notifying a Next Js Frontend about a Completed Process from a Backend?

### DISCUSSION THREAD

**Discussion by gustavostz** [vercel/next.js#45868] [Help]
## Situation:

I'm building a Next Js frontend that communicates with a Spring Boot backend through a Next Js API (BFF). The backend performs a lengthy process (5 minutes) and returns status code 202 to indicate that the process has been accepted and is in progress asynchronously. Now, I need to notify the Next Js API and frontend client when the process is completed. In this case, what is the best approach for this notification?

## Possibilities that I found:

 1. Firebase Cloud Message
 2. WebSocket
 3. Server-Sent Events (SSE)
 4. Short/Long Pooling

## What would be the best approach?

Due to the fact that cloud providers usually have short timeouts and costs are based on time of each operation, I wondered what the best approach would be to this scenario. 

### Analysing each possibility:

*Please feel free to correct any wrong analyses I made when comparing. And feel free to add more possibilities.*

#### Firebase Cloud Message:

 - Firebase Cloud Message seems like a good option as it will only notify the client when there's a message to deliver, and there won't be any problems with timeouts or costs.

 - However, it's usually used for push notifications, and the delivery of messages is not guaranteed, making it a less suitable option for this scenario.

 - Another positive thing about Cloud Message is that it is not necessary for the BFF to receive the message, because the Cloud Message will not tell the origin of the publisher.

#### WebSocket:

 - In this approach, the Next Js API establishes a persistent connection with the Spring Boot backend using WebSockets. The backend can then send a message to the API to notify it when the process is finished.

 - The biggest issue with this approach is the need to open a WebSocket connection with each client on the Next Js API, which is not so good for scalability as the first option.

 - Additionally, you have a bigger problem here, some Next Js hosting providers like [Vercel do not support WebSockets][1] as they are stateless and have a maximum execution duration, making it impossible to maintain a WebSocket connection.

*PS: I'm using Vercel, so it's not an available option.*

#### Server-Sent Events (SSE):

 - In this approach, the Next Js API establishes a persistent connection with the Spring Boot backend using SSE. The backend can then send an event to the API to notify it when the process is finished.

 - However, this approach relies on the same problem as WebSocket, as the Next JS API will have to maintain a connection open for each client.

 - Also, some hosting providers may not support SSE for the same reason as WebSocket.

#### Short/Long Pooling

 - In this approach, the Next Js frontend periodically sends a request to the Next Js API (BFF) to check the status of the process in the Spring Boot backend.

 - This is the least performant and cost-effective option as each request counts as a separate request and is charged by the cloud provider. And as an example of a performance problem, if you try to put a pooling of 60 seconds in an effort to avoid too many charges, but the application finishes the process in 63 seconds, you will have to wait 120 seconds for the result instead of 63 seconds.

 - Looking at the options, however, it seems to be the only viable option, because it will ensure delivery and is supported by all cloud providers.

## Questions

1. Are there any other possibilities for this scenario that I may have missed?
2. What is the best approach for this scenario and why?

  [1]: https://vercel.com/guides/do-vercel-serverless-functions-support-websocket-connections

**Comment by wwwdepot** [comment]
If you're using Firebase, consider their realtime db.

Just need to wire up the spring boot backend with it, frontend users will get a better experience, imo.

**Reply by gustavostz** [reply]
Good point. It's a possibility, but in this case, we have 2 drawbacks. The first is that real-time databases exist only in a single region, so if the product is global (which is the case) it will be less efficient. Another problem is related to security. It is generally not recommended to connect to a database directly from the client side as it can increase the risk of security vulnerabilities. Firebase has security rules that restrict access to the database, but sometimes it is a risk we cannot take.

**Comment by fuma-nama** [comment]
As an alternative to websocket, I use `ably` for real-time events
It works well with next.js. There’re also alternatives like pusher channels or Pubhub

However if you already using firebase for notifications, authentication or other services, I personally think Firebase’s real-time db is a better choice
